client/app/scripts/components/timeline-control.js

```diff
@@ -0,0 +1,251 @@
+
+    // If the slider value is less than 1s away from the right-end (current time),
+    // assume we meant the current time - this is important for the '... so far'
+    // ranges where the range of values changes over time.
+    if (millisecondsInPast < 1000) {
+      millisecondsInPast = 0;
+    }
+
+    this.setState({ millisecondsInPast });
+    this.debouncedUpdateTimestamp(millisecondsInPast);
+    this.props.startWebsocketTransition();
+  }
+
+  handleRangeOptionClick(rangeOption) {
+    this.setState({ rangeOptionSelected: rangeOption });
+
+    const rangeMilliseconds = this.getRangeMilliseconds(rangeOption);
+    if (this.state.millisecondsInPast > rangeMilliseconds) {
+      this.setState({ millisecondsInPast: rangeMilliseconds });
+      this.updateTimestamp(rangeMilliseconds);
+      this.props.startWebsocketTransition();
+    }
+  }
+
+  handleJumpToNowClick() {
+    this.setState({
+      showSliderPanel: false,
+      millisecondsInPast: 0,
+      rangeOptionSelected: sliderRanges.last1Hour,
+    });
+    this.updateTimestamp();
+    this.props.startWebsocketTransition();
+  }
+
+  handleTimestampClick() {
+    this.setState({ showSliderPanel: !this.state.showSliderPanel });
+  }
+
+  updateTimestamp(millisecondsInPast = 0) {
+    this.props.websocketQueryInPast(millisecondsInPast);
+    this.props.clickResumeUpdate();
+  }
+
+  getRangeMilliseconds(rangeOption = this.state.rangeOptionSelected) {
+    return moment().diff(rangeOption.getStart());
+  }
+
+  renderRangeOption(rangeOption) {
+    const handleClick = () => { this.handleRangeOptionClick(rangeOption); };
+    const selected = (this.state.rangeOptionSelected.label === rangeOption.label);
+    const className = classNames('option', { selected });
+
+    return (
+      <a key={rangeOption.label} className={className} onClick={handleClick}>
+        {rangeOption.label}
+      </a>
+    );
+  }
+
+  renderJumpToNowButton() {
+    return (
+      <a className="button jump-to-now" title="Jump to now" onClick={this.handleJumpToNowClick}>
+        <span className="fa fa-step-forward" />
+      </a>
+    );
+  }
+
+  renderTimelineSlider() {
+    const { millisecondsInPast } = this.state;
+    const rangeMilliseconds = this.getRangeMilliseconds();
+
+    return (
+      <Slider
+        onChange={this.handleSliderChange}
+        value={rangeMilliseconds - millisecondsInPast}
+        max={rangeMilliseconds}
+      />
+    );
+  }
+
+  render() {
+    const { websocketTransitioning, hasTimelineControl } = this.props;
+    const { showSliderPanel, millisecondsInPast } = this.state;
+    const isCurrent = (millisecondsInPast === 0);
+
+    // Don't render the timeline control if it's not explicitly enabled for this instance.
+    if (!hasTimelineControl) return null;
+
+    return (
+      <div className="timeline-control">
+        {showSliderPanel && <div className="slider-panel">
+          <span className="caption">Explore</span>
```

Comment:
Yeah, I might as well do that given that we're gonna cleanup the upper part of the component anyway :P